leader-election-impl/src/main/java/com/palantir/paxos/SqliteConnections.java

```diff
@@ -18,46 +18,54 @@
 
 import java.io.IOException;
 import java.nio.file.Path;
-import java.sql.Connection;
-import java.sql.SQLException;
-import java.util.function.Supplier;
+
+import javax.sql.DataSource;
 
 import org.apache.commons.io.FileUtils;
 import org.sqlite.SQLiteConfig;
 import org.sqlite.javax.SQLiteConnectionPoolDataSource;
 
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
 import com.palantir.logsafe.SafeArg;
 import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
 
 /**
  * This class is responsible for creating Sqlite connections to an instance.
  * There should be one instance per timelock.
  */
 public final class SqliteConnections {
     private static final String DEFAULT_SQLITE_DATABASE_NAME = "sqliteData.db";
+    private static final LoadingCache<Path, DataSource> DATA_SOURCE_LOADING_CACHE = Caffeine.newBuilder()
+            .build(SqliteConnections::buildDataSource);
 
     private SqliteConnections() {
         // no
     }
 
-    public static Supplier<Connection> createDefaultNamedSqliteDatabaseAtPath(Path path) {
+    public static DataSource getOrCreateDefaultDataSource(Path path) {
+        return DATA_SOURCE_LOADING_CACHE.get(path);
+    }
+
+    private static DataSource buildDataSource(Path path) {
         createDirectoryIfNotExists(path);
         String target = String.format("jdbc:sqlite:%s", path.resolve(DEFAULT_SQLITE_DATABASE_NAME).toString());
 
         SQLiteConfig config = new SQLiteConfig();
         config.setPragma(SQLiteConfig.Pragma.JOURNAL_MODE, SQLiteConfig.JournalMode.WAL.getValue());
-        config.setBusyTimeout(5000);
+        config.setPragma(SQLiteConfig.Pragma.LOCKING_MODE, SQLiteConfig.LockingMode.EXCLUSIVE.getValue());
+        config.setPragma(SQLiteConfig.Pragma.SYNCHRONOUS, SQLiteConfig.SynchronousMode.FULL.getValue());
+
         SQLiteConnectionPoolDataSource dataSource = new SQLiteConnectionPoolDataSource();
         dataSource.setUrl(target);
         dataSource.setConfig(config);
 
-        return () -> {
-            try {
-                return dataSource.getConnection();
-            } catch (SQLException e) {
-                throw new RuntimeException(e);
-            }
-        };
+        HikariConfig hikariConfig = new HikariConfig();
+        hikariConfig.setDataSource(dataSource);
+        hikariConfig.setMaximumPoolSize(1);
```

Comment:
can we have this be more than 1? Say 16?

Reply:
Like 1 just seems wrong.

Reply:
(and maybe remove the locking mode exclusive pragma...

Reply:
Like, just to be clear here: 1 means that you'll only have one query in flight to SQLite at a time, which seems strictly worse than shared locking.

Reply:
That's a good point, having more than 1 connection and removing exclusive locking means we can do parallel reads

Reply:
Yeah, in hindsight our tests probably overindexed on writes (which might be fair given that migration specifically has been causing us pain).